storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2288 @@ public class UnifiedLog {
+     * Get all segments beginning with the segment that includes "from" and ending with the segment
+     * that includes up to "to-1" or the end of the log (if to > logEndOffset).
+     */
+    public List<LogSegment> logSegments(long from, long to) {
+        synchronized (lock) {
+            return List.copyOf(localLog.segments().values(from, to));
+        }
+    }
+
+    public List<LogSegment> nonActiveLogSegmentsFrom(long from) {
+        synchronized (lock) {
+            return List.copyOf(localLog.segments().nonActiveLogSegmentsFrom(from));
+        }
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder logString = new StringBuilder();
+        logString.append("Log(dir=");
+        logString.append(dir());
+        topicId.ifPresent(id -> {
+            logString.append(", topicId=");
+            logString.append(id);
+        });
+        logString.append(", topic=");
+        logString.append(topicPartition().topic());
+        logString.append(", partition=");
+        logString.append(topicPartition().partition());
+        logString.append(", highWatermark=");
+        logString.append(highWatermark());
+        logString.append(", lastStableOffset=");
+        logString.append(lastStableOffset());
+        logString.append(", logStartOffset=");
+        logString.append(logStartOffset());
+        logString.append(", logEndOffset=");
+        logString.append(logEndOffset());
+        logString.append(")");
+        return logString.toString();
+    }
+
+    public void replaceSegments(List<LogSegment> newSegments, List<LogSegment> oldSegments) throws IOException {
+        synchronized (lock) {
+            localLog.checkIfMemoryMappedBufferClosed();
+            List<LogSegment> deletedSegments = LocalLog.replaceSegments(localLog.segments(), newSegments, oldSegments, dir(), topicPartition(),
+                    config(), scheduler(), logDirFailureChannel(), logIdent, false);
+            deleteProducerSnapshots(deletedSegments, true);
+        }
+    }
+
+    /**
+     * This function does not acquire Log.lock. The caller has to make sure log segments don't get deleted during
+     * this call, and also protects against calling this function on the same segment in parallel.
+     *
+     * <p>Currently, it is used by LogCleaner threads on log compact non-active segments only with LogCleanerManager's lock
+     * to ensure no other LogCleaner threads and retention thread can work on the same segment.
+     */
+    public Collection<Long> getFirstBatchTimestampForSegments(Collection<LogSegment> segments) {
+        return segments.stream().map(LogSegment::getFirstBatchTimestamp).toList();
+    }
+
+    /**
+     * Remove deleted log metrics
+     */
+    public void removeLogMetrics() {
+        metricNames.forEach(metricsGroup::removeMetric);
+        metricNames.clear();
+    }
+
+    private <T> T maybeHandleIOException(String msg, StorageAction<T, IOException> fun) throws KafkaStorageException {
```

Comment:
I open https://issues.apache.org/jira/browse/KAFKA-19076 to change `String msg` to `Supplier<String> msg`.